src/base1.rs

```diff
@@ -0,0 +1,1947 @@
+    /// Returns `true` if the entry is removed from the skip list.
+    pub fn is_removed(&self) -> bool {
+        self.node.is_removed()
+    }
+
+    /// Returns a reference to the key.
+    pub fn key(&self) -> &K {
+        &self.node.key
+    }
+
+    /// Returns a reference to the value.
+    pub fn value(&self) -> &V {
+        &self.node.value
+    }
+}
+
+impl<'a, 'g, K, V> Entry<'a, 'g, K, V>
+where
+    K: Ord + Send + 'static,
+    V: Send + 'static,
+{
+    /// Attempts to pin the entry with a reference count, ensuring that it
+    /// remains accessible even after the `Guard` is dropped.
+    ///
+    /// This method may return `None` if the reference count is already 0 and
+    /// the node has been queued for deletion.
+    pub fn pin(&self) -> Option<RefEntry<'a, K, V>> {
+        unsafe {
+            if self.node.increment() {
+                Some(RefEntry {
+                    parent: self.parent,
+                    node: &*(self.node as *const _),
+                })
+            } else {
+                None
+            }
+        }
+    }
+
+    /// Removes the entry from the skip list.
+    ///
+    /// Returns `true` if this call removed the entry and `false` if it was already removed.
+    pub fn remove(&self) -> bool {
+        // Try marking the tower.
+        if self.node.mark_tower() {
+            // Success - the entry is removed. Now decrement `len`.
+            self.parent.hot_data.len.fetch_sub(1, Ordering::Relaxed);
+
+            // Search for the key to unlink the node from the skip list.
+            self.parent
+                .search_bound(Bound::Included(&self.node.key), false, self.guard);
+
+            true
+        } else {
+            false
+        }
+    }
+}
+
+impl<'a, 'g, K, V> Clone for Entry<'a, 'g, K, V> {
+    fn clone(&self) -> Entry<'a, 'g, K, V> {
+        Entry {
+            parent: self.parent,
+            node: self.node,
+            guard: self.guard,
+        }
+    }
+}
+
+impl<'a, 'g, K, V> Entry<'a, 'g, K, V>
+where
+    K: Ord,
+{
+    /// Moves to the next entry in the skip list.
+    pub fn move_next(&mut self) -> bool {
+        match self.next() {
+            None => false,
+            Some(n) => {
+                *self = n;
+                true
+            }
+        }
+    }
+
+    /// Returns the next entry in the skip list.
+    pub fn next(&self) -> Option<Entry<'a, 'g, K, V>> {
+        let n = self.parent.next_node(
+            &self.node.tower,
+            Bound::Excluded(&self.node.key),
+            self.guard,
+        )?;
+        Some(Entry {
+            parent: self.parent,
+            node: n,
+            guard: self.guard,
+        })
+    }
+}
+
+/// A reference-counted entry in a skip list.
+///
+/// You *must* call `release` to free this type, otherwise the node will be
+/// leaked. This is because releasing the entry requires a `Guard`.
+pub struct RefEntry<'a, K: 'a, V: 'a> {
```

Comment:
If we implement "anonymous" pinning, then we'll be able to automatically release when dropping `RefEntry`:

Reply:
Are there any concrete plans for anonymous pinning? How would this work in practice?